### ISSUE TITLE
TopicPartitionWriter doesn't tolerate truncation of the partition by a broker

### ISSUE THREAD

**Issue by savulchik** [confluentinc/kafka-connect-hdfs#132]
You can get into the situation when the partition is truncated by the broker to the point after the last processed offset. When this happens you get:

> [2016-10-03 17:40:56,802] INFO Fetch offset 754769741 is out of range for partition bss.input-54, resetting offset (org.apache.kafka.clients.consumer.internals.Fetcher:585)
> [2016-10-03 17:41:23,675] INFO Ignoring stale out-of-order record in bss.input-54. Has offset 757195092 instead of expected offset 754769741 (io.confluent.connect.hdfs.TopicPartitionWriter:502)

After that TopicPartitionWriter begins to skip records indefinitely because the expected offset is already unreachable -- it is behind the earliest offset in the partition.

**Comment by rossoft** [comment]
We are having the same problem on a topic that has compaction enabled. As there are some offsets that are "skipped" by the compaction, TopicPartitionWriter gets stuck in this problem  as well.
I think the fix is as simple as removing the protection condition that skips record when record.kafkaOffset() != expectedOffset. That condition was added as a workaround for KAFKA-2894 as documented on source code:

```
    } else if (record.kafkaOffset() != expectedOffset) {
      // Currently it's possible to see stale data with the wrong offset after a rebalance when you
      // rewind, which we do since we manage our own offsets. See KAFKA-2894.
```

https://issues.apache.org/jira/browse/KAFKA-2894 is already fixed, so it should be safe to remove this workaround, just continuing if there is any stale record, as that just means a compaction / truncation in the topic. That will fix the problem we are facing.

**Comment by rossoft** [comment]
I think this PR will fix this.
https://github.com/confluentinc/kafka-connect-hdfs/pull/154

**Comment by savulchik** [comment]
@rossoft I suppose it should because we are no more stuck skipping records in case of truncation.

**Comment by savulchik** [comment]
@rossoft now I realize that it probably introduce other problems for my case, e.g. with offset counters. I believe that offset counter should be advanced forward with explicit notification in logs.

**Comment by rossoft** [comment]
Hi @savulchik, can you give me more details which problem related to offset counters introduces in your case? Thanks!!

**Comment by chenchik** [comment]
I'm still experiencing this issue using the docker image below as well as openshift.

docker image: https://hub.docker.com/r/confluentinc/cp-kafka-connect/

I run into this issue when trying to scale workers in order to try to push the most up to date data to hdfs so it doesn't fall too far behind the Kafka stream. I'm scaling workers simply by adding more "pods" in openshift. So I'm basically just adding replicas of the docker image and adding the same connector to those docker images/workers. 

I'm guessing this happens because one worker starts pushing to hdfs immediately, but the other worker still needs to boot up. So the second worker falls behind and the offset becomes much greater than it is expecting.

I have gotten this to work if I delete the topic directory in hdfs, but I don't want to do that every time. I could move the data to a separate directory and delete the original topic directory every time, but I would love to get it working properly.

It looks like the solution you posted in #154 isn't currently reflected on github:

https://github.com/confluentinc/kafka-connect-hdfs/blob/master/src/main/java/io/confluent/connect/hdfs/TopicPartitionWriter.java

where

`} else if (record.kafkaOffset() != expectedOffset) {`

should be:

`} else if (record.kafkaOffset() < expectedOffset) {`

I know that the docker image uses the source code below:

https://github.com/confluentinc/cp-docker-images/tree/3.2.x/java/src/main/java/io/confluent/admin/utils/cli

But I haven't been able to find the equivalent of the equality bug (< instead of !=) in there. 

Was this code change ever implemented for the docker image?

**Comment by ewencp** [comment]
This check should actually be safe to remove now anyway. I've filed #212 to do so.

**Comment by chenchik** [comment]
Would it be possible to update the docker image to pull the latest version with the new changes? 

I'm still getting print statements that shouldn't exist in the updated version:

`[2017-07-17 14:30:37,546] INFO Ignoring stale out-of-order record in topic-3-0. Has offset 13184146891 instead of expected offset 12667757341 (io.confluent.connect.hdfs.TopicPartitionWriter)`

/java/kafka-connect-hdfs/kafka-connect-hdfs-3.2.2.jar is still being used in the docker image but it looks like the version you updated is kafka-connect-hdfs-3.4.0.jar according to the Jenkins build mentioned in #212:

jenkins link:

https://jenkins.confluent.io/job/kafka-connect-hdfs-pr/99/io.confluent$kafka-connect-hdfs/

**Comment by ewencp** [comment]
The target was 3.4.x because 3.3.0 is about to come out and this isn't a regression. So the next release it would come out in is a 3.4.0. The docker images would be updated with that version when that release comes out.

**Comment by ewencp** [comment]
Also, I'm going to close this out now as it should be fixed by #212, but please reopen and provide logs if you encounter this even with that change included.
